### PR TITLE
action function calls for reward calculation

### PR DESCRIPTION
**Pull request by eesmerdag**
https://www.notion.so/figmentnetworks/tracking-stake-unstake-events-for-reward-calculation-28727405448743b6a450f87841474abc

### CODE REVIEW

File: near/action_function_calls.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+func ParseAction(trx model.Transaction, account, actionType, validator string) (amt *big.Int, err error) {
+	amt = new(big.Int)
+	var receiptList []ReceiptsOutcome
+	var outcome TransactionOutcome
+	err = json.Unmarshal(trx.Outcome, &outcome)
+	if err != nil {
+		return amt, err
+	}
+	if outcome.Outcome.Status.SuccessReceiptID == nil {
+		return amt, errors.New("error outcome.status.SuccessReceiptID nil")
```
Comment (sosedoff) [inline near/action_function_calls.go:41]: i think the outcome status may include the actual error?

File: near/action_function_calls.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+	}
+
+	err = json.Unmarshal(trx.Receipt, &receiptList)
+	if err != nil {
+		return amt, err
+	}
+
+	// log type 1: where receiver is the pool
+	// seen as an example: 9MyRhDWxHEtdEsqscnx9vfvaRpNDSWS2QhMd7kVPveHx
+	if strings.Contains(trx.Receiver, "pool") && !(trx.Receiver == validator) {
```
Comment (sosedoff) [inline near/action_function_calls.go:51]: `trx.Receiver != validator` ?

File: near/action_function_calls.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+					return getAmountFromLog(logV, msg)
+				}
+			}
+		}
+	}
+
+	return amt, nil
+}
+
+func getAmountFromLog(log, amountMsg string) (n *big.Int, err error) {
```
Comment (sosedoff) [inline near/action_function_calls.go:96]: can you add tests for anything that does content transformations?

File: store/queries/transactions_confirm_actions.sql [hunk 4]
```diff
@@ -0,0 +1,12 @@
+SELECT
+  *
+FROM
+  (SELECT *, arr.item_object -> 'data' -> 'method_name' AS method_name
```
Comment (sosedoff) [inline store/queries/transactions_confirm_actions.sql:4]: Use `object->'attr'` notation here, without any extra spacing.

File: store/queries/transactions_confirm_actions.sql [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+      FROM transactions,
+           jsonb_array_elements(actions) WITH ORDINALITY arr(item_object, position)
+      WHERE arr.item_object ->>'type' = 'FunctionCall') AS tbl_trxs_confirm
```
Comment (sosedoff) [inline store/queries/transactions_confirm_actions.sql:7]: There does not seem to be any limit on this subquery, is that intentional? 
You can also clean it up with a CTE:

```
WITH tbl_trxs_confirm AS (
   ...
)
SELECT * FROM tbl_trxs_confirm ....
```

Another point is that we have a lot of transactions, so any unoptimized query will have a huge performance hit on the DB.

File: pipeline/task_parser.go [hunk 6]
```diff
@@ -71,40 +74,167 @@ func (t ParserTask) Run(ctx context.Context, payload *Payload) error {
...
+										}
+									}
+								}
+							}
+
+							// handle if there is stake/unstake confirm action for delegator
+							_, ok = confirmActions[de.AccountID]
+							if ok {
+								for _, act := range confirmActions[de.AccountID] {
+									if act.Validator == de.ValidatorID {
```
Comment (sosedoff) [inline pipeline/task_parser.go:221]: This conditional is duplicate: `act.Validator == de.ValidatorID`.

File: pipeline/task_parser.go [hunk 7]
```diff
@@ -71,40 +74,167 @@ func (t ParserTask) Run(ctx context.Context, payload *Payload) error {
...
+										actDataEncoded, err := json.Marshal(act.Data)
+										if err != nil {
+											return err
+										}
+										err = json.Unmarshal(actDataEncoded, &actInfo)
+										if err != nil {
+											return err
+										}
+
+										switch actInfo.MethodName {
```
Comment (sosedoff) [inline pipeline/task_parser.go:169]: Can we somehow flatten this method body a bit? There's way too much nesting going on. Maybe extract some functionality into smaller methods.

**Review by sosedoff — COMMENTED** [review]
Will do another pass, but here's the initial CR

File: pipeline/task_parser.go (hunk 7 above)
Comment (eesmerdag) [reply pipeline/task_parser.go]: https://github.com/figment-networks/near-indexer/pull/49/commits/37e890fd4cc7325a7b7a1ce89eb6e5781cc2d072

File: near/action_function_calls.go (hunk 3 above)
Comment (eesmerdag) [reply near/action_function_calls.go:97]: https://github.com/figment-networks/near-indexer/pull/49/commits/bb3f83d26e0895b0110d26f346df332f39928ad9

File: near/action_function_calls.go [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+	}
+
+	err = json.Unmarshal(trx.Receipt, &receiptList)
+	if err != nil {
+		return amt, err
+	}
+
+	// log type 1: where receiver is the pool
+	// seen as an example: 9MyRhDWxHEtdEsqscnx9vfvaRpNDSWS2QhMd7kVPveHx
+	if strings.Contains(trx.Receiver, "pool") && trx.Receiver != validator {
```
[the remark this answers is not in the document]
Comment (eesmerdag) [reply near/action_function_calls.go]: https://github.com/figment-networks/near-indexer/pull/49/commits/9ce0075957fb718a1e8e544974a5aa76963d7d91

File: near/action_function_calls.go [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+	ActionTypeUnstake = "unstake"
+	ActionTypeStake   = "stake"
+)
+
+// 	we only parse needed log message types, there is more than handled in here.
+//  ParseAction to get unstaked or staked amount
+func ParseAction(trx model.Transaction, account, actionType, validator string) (amt *big.Int, err error) {
+	amt = new(big.Int)
+	var receiptList []ReceiptsOutcome
+	var outcome TransactionOutcome
```
[the remark this answers is not in the document]
Comment (eesmerdag) [reply near/action_function_calls.go:35]: https://github.com/figment-networks/near-indexer/pull/49/commits/2f6925b0a3e3f02541711f630832b57435b827ac

File: near/action_function_calls.go [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+	ActionFunctionCallWithdrawFromStakingPool    = "withdraw_from_staking_pool"
+	ActionFunctionCallWithdrawAllFromStakingPool = "withdraw_all_from_staking_pool"
+
+	ActionTypeUnstake = "unstake"
+	ActionTypeStake   = "stake"
+)
+
+// 	we only parse needed log message types, there is more than handled in here.
+//  ParseAction to get unstaked or staked amount
+func ParseAction(trx model.Transaction, account, actionType, validator string) (amt *big.Int, err error) {
```
[the remark this answers is not in the document]
Comment (eesmerdag) [reply near/action_function_calls.go:32]: why do we need pointer receiver here? the function does not update trx but just getting related amount.

**Comment by sosedoff** [comment]
@eesmerdag could you rebase this with master?

**Comment by eesmerdag** [comment]
> @eesmerdag could you rebase this with master?

applied this commit sir. now lgty?
https://github.com/figment-networks/near-indexer/pull/49/commits/181176f55259a505877888d49b1a6ee353025b15

**Comment by sosedoff** [comment]
I was testing all this stuff locally, but didnt want to break any of the code you wrote.

**Pull request merged**